controllers/s3utils.go

```diff
@@ -0,0 +1,404 @@
+	// does not support concurrent writers.
+	s3Uploader := s3manager.NewUploaderWithClient(s3Client)
+	s3Downloader := s3manager.NewDownloaderWithClient(s3Client)
+	s3Conn := &s3Connection{
+		session:    s3Session,
+		client:     s3Client,
+		uploader:   s3Uploader,
+		downloader: s3Downloader,
+		endpoint:   s3Endpoint,
+		callerTag:  callerTag,
+	}
+	objectStoreConnections[s3Endpoint] = s3Conn
+
+	return s3Conn, nil
+}
+
+func getS3Secret(ctx context.Context, r client.Reader,
+	s3SecretName types.NamespacedName) (
+	s3AccessID, s3SecretAccessKey []byte, err error) {
+	secret := corev1.Secret{}
+	if err := r.Get(ctx, s3SecretName, &secret); err != nil {
+		log.Infof("Failed to get secret %v", s3SecretName)
+
+		return nil, nil, errorswrapper.Wrapf(err,
+			"Failed to get secret %v", s3SecretName)
+	}
+
+	s3AccessID = secret.Data["AWS_ACCESS_KEY_ID"]
+	s3SecretAccessKey = secret.Data["AWS_SECRET_ACCESS_KEY"]
+
+	return
+}
+
+// Create a bucket; don't return error if the bucket exists already
+func (s *s3Connection) createBucket(bucket string) error {
+	if bucket == "" {
+		return fmt.Errorf("error: unspecified bucket name for "+
+			"endpoint %v tag %v",
+			s.endpoint, s.callerTag)
+	}
+
+	_, err := s.client.CreateBucket(
+		&s3.CreateBucketInput{Bucket: &bucket})
+	if err != nil {
+		if !errorswrapper.As(err, s3.ErrCodeBucketAlreadyOwnedByYou) {
+			return errorswrapper.Wrapf(err, "Failed to create bucket %s",
+				bucket)
+		}
+	}
+
+	return nil
+}
+
+// A convenience method
+// - OK to call UploadPV() concurrently from multiple goroutines safely.
+func (s *s3Connection) uploadPV(bucket string, pvKeySuffix string,
+	pv corev1.PersistentVolume) error {
+	return s.uploadTypedObject(bucket, pvKeySuffix /* key suffix */, pv)
+}
+
+// Upload to the given bucket the given uploadContent with a key of
+// <objectType/keySuffix>, where objectType is the type of the uploadContent
+// parameter
+// - OK to call UploadTypedObject() concurrently from multiple goroutines safely.
+func (s *s3Connection) uploadTypedObject(bucket string, keySuffix string,
+	uploadContent interface{}) error {
+	keyPrefix := reflect.TypeOf(uploadContent).String() + "/"
+	key := keyPrefix + keySuffix
+
+	return s.uploadObject(bucket, key, uploadContent)
+}
+
+// Upload the given object to the given bucket with the given key
+// - OK to call UploadObject() concurrently from multiple goroutines safely.
+// - Upload may fail due to many reasons: RequestError (connection error),
+//	 NoSuchBucket, NoSuchKey, InvalidParameter (e.g., empty key), etc.
+// - Multiple consecutive forward slashes in the key are sqaushed to
+//	 a single forward slash, for each such occurrence
+// - Any formatting changes to this method should also be reflected in the
+//	 downloadObject() method
+func (s *s3Connection) uploadObject(bucket string, key string,
+	uploadContent interface{}) error {
+	encodedUploadContent := &bytes.Buffer{}
+
+	gzWriter := gzip.NewWriter(encodedUploadContent)
```

Comment:
Probably add `defer gzWriter.Close()` instead of having it in line 233.

Reply:
Given that I want to return an err if close fails,  I need to defer a func() to change the return value.  Is that what you mean?